File: 1818-maximum-score-from-removing-substrings/maximum-score-from-removing-substrings.java
class Solution {
    public int maximumGain(String s, int x, int y) {
        StringBuilder sb = new StringBuilder(s);
        char first = x > y ? 'a' : 'b';
        char second = x > y ? 'b' : 'a';
        int high = Math.max(x, y);
        int low = Math.min(x, y);
        
        int score = removePatterns(sb, first, second, high);
        score += removePatterns(sb, second, first, low);
        
        return score;
    }
    
    private int removePatterns(StringBuilder sb, char first, char second, int points) {
        int score = 0;
        StringBuilder stack = new StringBuilder();
        
        for (int i = 0; i < sb.length(); i++) {
            if (!stack.isEmpty() && sb.charAt(i) == second && stack.charAt(stack.length() - 1) == first) {
                stack.setLength(stack.length() - 1);
                score += points;
            } else {
                stack.append(sb.charAt(i));
            }
        }
        
        sb.setLength(0);
        sb.append(stack);
        return score;
    }
}

File: 1818-maximum-score-from-removing-substrings/maximum-score-from-removing-substrings.c
int max(int a, int b) {
    return (a > b) ? a : b;
}

int min(int a, int b) {
    return (a < b) ? a : b;
}

int maximumGain(char* s, int x, int y) {
    int len = strlen(s);
    int score = 0;
    char *stack = (char*)malloc(len * sizeof(char));
    int top = -1;
    char first = x > y ? 'a' : 'b';
    char second = x > y ? 'b' : 'a';
    int high = max(x, y);
    int low = min(x, y);

    // First pass
    for (int i = 0; i < len; i++) {
        if (top >= 0 && s[i] == second && stack[top] == first) {
            top--;
            score += high;
        } else {
            stack[++top] = s[i];
        }
    }

    // Reset stack for second pass
    int tempLen = top + 1;
    char *temp = (char*)malloc(tempLen * sizeof(char));
    memcpy(temp, stack, tempLen * sizeof(char));
    top = -1;

    // Second pass
    for (int i = 0; i < tempLen; i++) {
        if (top >= 0 && temp[i] == first && stack[top] == second) {
            top--;
            score += low;
        } else {
            stack[++top] = temp[i];
        }
    }

    free(stack);
    free(temp);
    return score;
}